src/Core/Aetherling/Passes/ThroughputModifications.hs

```diff
@@ -1,339 +1,508 @@
--- made full throughput must consume the rest of throuhgputMult and result in a
--- new pxPerClock that cleanly divides into that dimension.
--- Stated rigorously:
---    (first not full throuhgput dim) % (throuhgMult) / (product of inner px
---     per clock increases) == 0
-attemptSpeedUp throughMult (LineBuffer p w img t bc) =
+-- LEAF, NON-MODIFIABLE RATE 
+-- can't change rate, can't speed up child ops, so just wrap in a map to
+-- parallelize
+-- ASSUMPTION: the user has specified a type for these ops and passes will not
+-- change the type because that would change the semantics of the
+-- program. For example, one could speed up an Add T_Int my making it a
+-- Add $ T_Array 2 T_Int, but that would change the program's meaning.
+attemptSpeedUp requestedMult op@(Add _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Sub _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Mul _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Div _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Max _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Min _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Ashr _ _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Shl _ _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Abs _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Not _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(And _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Or  _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(XOr _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Eq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Neq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Lt = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Leq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Gt = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Geq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(LUT _) = (MapOp requestedMult op, requestedMult)
+
+-- Must map over these because making a single memory wider is a more
+-- expensive operation than banking it (making multiple copies)
+-- NOTE: unclear what expensive means here, just my understanding from
+-- talking with Spatial team, Ross, and other hardware people. I should
+-- figure it out.
+-- maping over a MemRead/MemWrite is banking it, and wrapping
+-- the type with an array is making a single memory read/write more
+-- per clock.
+attemptSpeedUp requestedMult op@(MemRead _) = (MapOp requestedMult op, requestedMult)
```

Comment:
#braindump as this review caused me to think about how we treat memories.

TLDR 1; I think memRd and memWr should be leaf modifiable rate operators given their current definition.

TLDR 2; in the comment on line 102 above, "banking" should be replaced by "gathering".

=====================

We currently consider MemRead and MemWrite to be stateful ops.  (as given in Operations/Properties.hs). These ops are stateful because they are currently defined to internally contain an address generator.  And we know that we cannot throw a map around a stateful op without changing the program's behavior (recall the linebuffer example -- each copy of the linebuffer now gets a different substream of non-consecutive elements) 

It's possible, and perhaps preferable to think of a memRd unit as two units that are composed via a composeSeq.  (1) A stateful address generator that emits a stream of addresses of a specified length (for simplicity, let's consider this stream a bunch of integers: 0 1 2 3 4..., or if there was some form of stride functionality in the address generator 0 4 8 12 16 ...). (2) And a stateless unit MemAccess that actually accesses memory at the incoming address, emitting the value at that address.  So now consider the following:

KayvonMemRead  == AddrGen N 1 |>>=| MemAccess   

(where AddrGen N 1 is an address generator generating N elements from throughput 1 and MemAccess is a stateless memory access unit.)

Now let's consider what speedup k KayvonMemRead should do:

Let's consider how to speedup AddrGen N 1.  

Option 1: If I make it an op that speeds up itself, I'd just change it to: AddrGen N k.
I'd expect the semantics are for it to emit vectors of addresses. For k=2 [0 1] [2 3] [3 4] etc.  Since the addresses are continuous, executing the subsequent memory access just needs a wider memory system.  

Alternative 2: Consider throwing a map around AddrGen N 1 to speed it up: map k AddrGen N 1 would yield k concurrent streams of size N, which is clearly wrong.  That's k times more elements being generated.

Alternative 3: Throwing a map around the op, and also dividing the resulting stream length is interesting to consider: 
speedup k AddrGen N 1 = map k AddrGen (N/k) 1, but it yields the following pattern for k=2:
    [0 2] [1 3] [2 4]...

This results in N total elements in a streams of length N/2, with element type int[2]. One might think that regardless of whether I choose option 1 or alternative 3 in how to speedup address generation, the stateless memAccess could be sped up as (map k MemAccess).  However, this is not true.

To me (map k MemAccess)  models *k* copies of a basic memory access unit.  The meaning here is that the HW could make k accesses to k independent addresses at a time, since there is nothing known about the addresses being provided as input.  The resulting memory itself would have to be able to arbitrate these k simultaneous requests.  

However, with knowledge of the incoming stream, very different, and far more efficient, implementations of map k MemAccess could be built. For example, if it was known that the address stream was that produced by Option 1 above, then the memory need only handle one bulk access at a time, with that access being to an array of consecutive values.  Let's call this parallel memory access operator MemAccessArray k.

Therefore...
map k MemAccess would be implemented as a special unit MemAccessArray k *if the incoming stream had consecutive addresses in each token*.  This would be implemented via a banked SRAM, or just a wider memory port.

Otherwise...
map k MemAccess would be equivalent to MemAccessGather k, a much more complex hardware unit.

So in other words, I feel like memRd needs to be a leaf modifiable rate operator.  Or if it was expressed in terms as a composeSeq of my hypothetical AddrGen and MemAccess/MemAccesArray/MemAccessGather ops, then all these ops need to be modifiable rate, and there's the equivalence:

memRd k = AddrGen N k |>>=| MemArray Rd k

Reply:
This will require a significant simulator change. I will delay this change until later.

Reply:
See https://github.com/David-Durst/aetherlingHaskellIR/issues/17 for progress on this comment.